<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

In [3]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [6]:
query_3_1 = f''' 
select 
    count(v.id) -- выбираем агрегацию столбца для вывода 
from vacancies as v -- из таблицы вакансий с алиасом
'''

In [7]:
df3_1 = pd.read_sql_query(query_3_1, connection) # создаём датафрейм из запроса
df3_1 # выводим датафрейм

C:\Users\usuario\AppData\Local\Temp\ipykernel_13628\3662297495.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df3_1 = pd.read_sql_query(query_3_1, connection) #создаём датафрейм из запроса


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [8]:
query_3_2 = f''' 
select 
    count(e.id) -- выбираем агрегацию столбца для вывода 
from employers as e -- из таблицы вакансий с алиасом
'''

In [9]:
df3_2 = pd.read_sql_query(query_3_2, connection) # создаём датафрейм из запроса
df3_2 # выводим датафрейм

C:\Users\usuario\AppData\Local\Temp\ipykernel_13628\1586657423.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df3_2 = pd.read_sql_query(query_3_2, connection) #создаём датафрейм из запроса


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [10]:
query_3_3 = f''' 
select 
    count(a.id) -- выбираем агрегацию столбца для вывода
from areas as a -- из таблицы вакансий с алиасом
'''

In [11]:
df3_3 = pd.read_sql_query(query_3_3, connection) # создаём датафрейм из запроса
df3_3 # выводим датафрейм

C:\Users\usuario\AppData\Local\Temp\ipykernel_13628\1503436232.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df3_3 = pd.read_sql_query(query_3_3, connection) #создаём датафрейм из запроса


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [12]:
query_3_4 = f''' 
select 
    count(i.id) -- выбираем агрегацию столбца для вывода
from industries as i -- из таблицы вакансий с алиасом
'''

In [13]:
df3_4 = pd.read_sql_query(query_3_4, connection) # создаём датафрейм из запроса
df3_4 # выводим датафрейм

C:\Users\usuario\AppData\Local\Temp\ipykernel_13628\4089142195.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df3_4 = pd.read_sql_query(query_3_4, connection) #создаём датафрейм из запроса


count
0    294

***

Предварительный анализ имеющихся у нас данных показал, что у нас есть самое большое число вакансий (49197), предлагаемые 23501 работодателями, в 1362 х регионах и 294х сферах деятельности. 

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [14]:
query_4_1 = f''' 
select 
    a.name as area,
    count(v.id) as cnt -- выбираем столбцы для вывода, наименование регионов и соответсвующее им количество вакансий
from vacancies as v -- из таблицы вакансий с алиасом v
join areas as a on v.area_id = a.id -- присоединяя таблицу регионов с алиасом a
group by a.name -- группируем по регионам
order by 2 desc -- упорядочиваем по количеству вакансий в порядке убывания
'''

In [15]:
df4_1 = pd.read_sql_query(query_4_1, connection) # создаём датафрейм из запроса
df4_1 # выводим датафрейм

C:\Users\usuario\AppData\Local\Temp\ipykernel_13628\116653629.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df4_1 = pd.read_sql_query(query_4_1, connection) #создаём датафрейм из запроса


area   cnt
0             Москва  5333
1    Санкт-Петербург  2851
2              Минск  2112
3        Новосибирск  2006
4             Алматы  1892
..               ...   ...
764       Тарко-Сале     1
765    Новоаннинский     1
766            Бирск     1
767           Сасово     1
768             Анна     1

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [16]:
query_4_2 = f''' 
select 
    count(v.id) cnt -- вбираем агрегацию счётчика для столбца id вакансий
from vacancies v -- из таблицы вакансий с алиасом v
where v.salary_from is not null or v.salary_to is not null -- задаём условие фильтрации запроса
'''

In [17]:
df4_2 = pd.read_sql_query(query_4_2, connection) # создаём датафрейм из запроса
df4_2 # выводим датафрейм

C:\Users\usuario\AppData\Local\Temp\ipykernel_13628\359678729.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df4_2 = pd.read_sql_query(query_4_2, connection) #создаём датафрейм из запроса


cnt
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [18]:
query_4_3 = f''' 
select 
    avg(v.salary_from) lower_salary,
    avg(v.salary_to) upper_salary -- выбираем агрегацию среднего значения для столбцов с верхним и нижним порогом зарплаты и задаём им алиасы
from vacancies v -- из таблицы вакансий с алиасом v
'''

In [19]:
df4_3 = pd.read_sql_query(query_4_3, connection) # создаём датафрейм из запроса
round(df4_3) # выводим датафрейм

C:\Users\usuario\AppData\Local\Temp\ipykernel_13628\4115627050.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df4_3 = pd.read_sql_query(query_4_3, connection) # создаём датафрейм из запроса


lower_salary  upper_salary
0       71065.0      110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [43]:
query_4_4 = f''' 
select 
    v.schedule "Тип графика",
    v.employment "Тип Трудоустройства",
    count(v.id) cnt -- выбираем столбцы типа графика и типа трудоцстройства, а так же агрегацию счётчика id вакансий и задаём соотв. алиасы
from vacancies v -- из таблицы вакансий
group by v.schedule, v.employment -- группируем по типам графика и трудоустройства
order by count(v.id) desc -- упорядочиваем по количеству вакансий в порядке убывания
'''

In [44]:
df4_4 = pd.read_sql_query(query_4_4, connection) # создаём датафрейм из запроса
df4_4 # выводим датафрейм

C:\Users\usuario\AppData\Local\Temp\ipykernel_13628\3912017390.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df4_4 = pd.read_sql_query(query_4_4, connection) # создаём датафрейм из запроса


Тип графика  Тип Трудоустройства    cnt
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [22]:
query_4_5 = f''' 
select 
    v.experience,
    count(v.id) cnt -- выбираем столбец опыта и агрегацию счётчика вакансий
from vacancies v -- из таблицы вакансий
group by v.experience -- группируем по типу опыта
order by count(v.id) -- упорядочиваем по количеству вакансий
'''

In [23]:
df4_5 = pd.read_sql_query(query_4_5, connection) # создаём датафрейм из запроса
df4_5 # выводим датафрейм

C:\Users\usuario\AppData\Local\Temp\ipykernel_13628\1608415065.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df4_5 = pd.read_sql_query(query_4_5, connection) # создаём датафрейм из запроса


experience    cnt
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

Мы выяснили как зависит количество вакансий от регионов, в больших городах предлагается больше вакансий, болше всего их в Москве и Санкт-Петербурге. Для 24073 вакансий заполненно хотя бы одно из 2х типов предложения оплаты, а это меньше 50% от всего количества вакансий. Так же мы выяснили средние значения минимального предложения оплаты и среднее значение максимального предложения оплаты. Мы узнали количество вакансий в зависимости от всевозможных сочетаний Типа графика и Типа трудоцстройства и выяснили, что в большинстве случаев работодатели предпочитают полную занятость и полный день, а реже всего сменный график и проектную работу. Ну и в конце мы выяснили зависимость количества вакансий от опыта работы, и наиболее запрашиваемым является опыт от года до 3х лет, затем идёт опыт от 2х лет до 6ти. Замечу, что наименьшее количество вакансий приходится, на тех у кого опыт более 6ти лет а не на тех у кого его(опыта) совсем нет.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [24]:
query_5_1 = f''' 
select 
    e.name,
    count(v.id) cnt -- выбираем столбец наименования работодателя и агрегацию счётчика вакансий
from vacancies v -- из таблицы вакансий
join employers e on v.employer_id = e.id -- присоединяя таблицу работодателей по ключу id
group by e.name -- группируем по наименованию работодателей
order by count(v.id) desc -- упорядочиваем по количеству вакансий в порядке убывания
limit 5 -- задаём предел вывода 5, поскольку нам интересны первй и пятый результаты по количеству вакансий
'''

In [25]:
df5_1 = pd.read_sql_query(query_5_1, connection) # создаём датафрейм из запроса
df5_1 # выводим датафрейм

C:\Users\usuario\AppData\Local\Temp\ipykernel_13628\579913344.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df5_1 = pd.read_sql_query(query_5_1, connection) # создаём датафрейм из запроса


name   cnt
0         Яндекс  1933
1     Ростелеком   491
2       Тинькофф   444
3           СБЕР   428
4  Газпром нефть   331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [26]:
query_5_2 = f''' 
select 
    a.name,
    count(distinct e.id) cnt_e,
    count(distinct v.id) cnt_v -- выбираем для вывода наименования регионов, а так же агрегации счётчиков для работодателей и вакансий считая только уникальные значения
from areas a -- из таблицы регионов с алиасом а
left join employers e on a.id = e.area -- присоединяя таблицу работотаделей по ключу, сохраняя все данные левой таблицы,т.е. регионов
left join vacancies v on a.id = v.area_id -- присоединяя таблицу вакансий по ключу, сохраняя все данные левой таблицы
where v.id is null -- задаём условие для выявления отсутсвия вакансий в регионе
group by a.id -- группируем по id региона
order by 2 desc -- упорядочиваем по количеству работодателей в порядке убывания
'''

In [27]:
df5_2 = pd.read_sql_query(query_5_2, connection) # создаём датафрейм из запроса
df5_2 # выводим датафрейм

C:\Users\usuario\AppData\Local\Temp\ipykernel_13628\3963928891.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df5_2 = pd.read_sql_query(query_5_2, connection) # создаём датафрейм из запроса


name  cnt_e  cnt_v
0                Россия    410      0
1             Казахстан    207      0
2    Московская область     75      0
3    Краснодарский край     19      0
4    Ростовская область     18      0
..                  ...    ...    ...
588             Скидель      0      0
589          Белоозерск      0      0
590           Орловский      0      0
591          Павловская      0      0
592             Таксимо      0      0

[593 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [28]:
query_5_3 = f''' 
select 
    e.name,
    count(distinct v.area_id) cnt_a -- выбираем для вывода столбец наименования работодателя и агрегацию счётчика регионов по id региона
from vacancies v -- из таблицы вакансий с алиасом v
join employers e on e.id = v.employer_id -- присоединяя таблицу работодателей по ключу с алиасом e
group by e.name -- группируем по наименованию работодателя
order by 2 desc -- упорядочиваем по количеству регионов в порядке убывания
'''

In [29]:
df5_3 = pd.read_sql_query(query_5_3, connection) # создаём датафрейм из запроса
df5_3 # выводим датафрейм

C:\Users\usuario\AppData\Local\Temp\ipykernel_13628\4098774629.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df5_3 = pd.read_sql_query(query_5_3, connection) # создаём датафрейм из запроса


name  cnt_a
0                      Яндекс    181
1                  Ростелеком    152
2                  Спецремонт    116
3      Поляков Денис Иванович     88
4                   ООО ЕФИН      71
...                       ...    ...
14761                  UniSol      1
14762            UNISTORY LLC      1
14763                   UNIT6      1
14764     United Distribution      1
14765              UNITEL LLC      1

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [30]:
query_5_4 = f''' 
select 
    count(e.id) cnt_e -- выбираем для вывода агрегацию счётчика для id работодателя
from employers e -- из таблицы работодателей
left join employers_industries ei on e.id = ei.employer_id -- присоединяя сводную таблицу работодателей и их сфер деятельности по ключу 
where ei.industry_id is null -- отбираем тех работодателей у кого сфера не указанна
'''

In [31]:
df5_4 = pd.read_sql_query(query_5_4, connection) # создаём датафрейм из запроса
df5_4 # выводим датафрейм

C:\Users\usuario\AppData\Local\Temp\ipykernel_13628\3918799020.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df5_4 = pd.read_sql_query(query_5_4, connection) # создаём датафрейм из запроса


cnt_e
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [32]:
query_5_5 = f''' 
select 
    e.name -- выбираем для вывода столбец с наименованиями работодателей
from employers e -- из таблицы работодателей с алиасом e
left join employers_industries ei on e.id = ei.employer_id -- присоединяя сводную таблицу работодателей и их сфер деятельности, сохраняя все данные таблицы работодателей
group by e.name -- группируем по наименованию работодателей
having count(ei.industry_id) = 4 -- оставляем в выборке только тех работодателей у кого насчитается 4 сферы деятельности
order by e.name -- упорядочиваем по наименованию работодателя
limit 3 -- выбираем к выводу только первых 3х
'''

In [33]:
df5_5 = pd.read_sql_query(query_5_5, connection)  # создаём датафрейм из запроса
df5_5 # выводим датафрейм

C:\Users\usuario\AppData\Local\Temp\ipykernel_13628\386902789.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df5_5 = pd.read_sql_query(query_5_5, connection)  # создаём датафрейм из запроса


name
0  101 Интернет
1      21vek.by
2          2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [34]:
query_5_6 = f''' 
select 
    count(e.id) -- выбираем для вывода количества работодателей агрегацию счётчика по id работодателя
from employers_industries ei -- к сводной таблице присоединим
join employers e on e.id = ei.employer_id -- таблицу с работодателями по соотв. ключу
join industries i on i.id = ei.industry_id -- и таблицу со сферами деятельности по соотв. ключу
where i.name = 'Разработка программного обеспечения' -- полученное соединение таблиц отфильтруем по условию
'''

In [35]:
df5_6 = pd.read_sql_query(query_5_6, connection) # создаём датафрейм из запроса
df5_6 # выводим датафрейм

C:\Users\usuario\AppData\Local\Temp\ipykernel_13628\4267767211.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df5_6 = pd.read_sql_query(query_5_6, connection) # создаём датафрейм из запроса


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [37]:
import requests # Импортируем библиотеку requests
from bs4 import BeautifulSoup # Импортируем библиотеку BeautifulSoup
url = 'https://ru.wikipedia.org/wiki/Города-миллионеры_России' # Определяем адрес страницы
response = requests.get(url) # Выполняем GET-запрос, содержимое ответа присваивается переменной response
page = BeautifulSoup(response.text, 'html.parser') # Создаём объект BeautifulSoup, указывая html-парсер
cities = (str(page.find('table', class_='standard'))) # Получаем тег table, отображающийся на вкладке браузера и записываем данные в переменную
cities = cities.split('title="')[3:] # разделяем по данные в срезе от 3его
million_cities = [] # задаём пустой список для заполнения
for i in cities: # задаём цикл в котором в каждой итерации будем вылавливать город-миллионик и добавлять его в пустой список
    million_cities.append(i.split('">')[0]) # отделяем название городов от лишних символов и добавляем в список
million_cities = tuple(million_cities) # превращаем список в кортеж для работы с ним в запросе
print(million_cities) # выводим результат на экран для визуальной проверки

('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Красноярск', 'Нижний Новгород', 'Челябинск', 'Уфа', 'Самара', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград')


In [46]:
query_5_7 = f''' 
(select 
    a.name,
    count(v.id) cnt -- выбираем для вывода столбец с наименованиями регионов и количество вакансий
from areas a -- из таблицы регионов с алиасом а
join vacancies v on a.id = v.area_id -- присоеднияя таблицу вакансий с алиасом v по ключу
join employers e on e.id = v.employer_id -- присоединяя таблицу работодателей с алиасом е по ключу
where e.name = 'Яндекс' -- задаём условие отбора по наименованию работодателя
group by a.name -- группируем по наименованию регионов
having a.name in {million_cities} -- задаём условие отбора(фильтрации) из кортежа с городами-миллиониками
)
union -- добавляем к первому запросу строку Total и сумму второго столбца
select
    'Total', -- выбираем название строки
    sum(cnt) -- добавляем агрегацию суммы столбца cnt
from (
select -- из созданной таблицы первого запроса
    a.name,
    count(v.id) cnt
from areas a
join vacancies v on a.id = v.area_id
join employers e on e.id = v.employer_id
where e.name = 'Яндекс'
group by a.name
having a.name in {million_cities}
) cntr
order by 2 -- общий полученный результат упорядочиваем по возрастанию количества
'''

In [47]:
df5_7 = pd.read_sql_query(query_5_7, connection) # создаём датафрейм из запроса
df5_7 # выводим датафрейм

C:\Users\usuario\AppData\Local\Temp\ipykernel_13628\2852167293.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df5_7 = pd.read_sql_query(query_5_7, connection) # создаём датафрейм из запроса


name    cnt
0              Омск   21.0
1         Челябинск   22.0
2        Красноярск   23.0
3         Волгоград   24.0
4             Пермь   25.0
5            Казань   25.0
6    Ростов-на-Дону   25.0
7               Уфа   26.0
8            Самара   26.0
9         Краснодар   30.0
10          Воронеж   32.0
11      Новосибирск   35.0
12  Нижний Новгород   36.0
13     Екатеринбург   39.0
14  Санкт-Петербург   42.0
15           Москва   54.0
16            Total  485.0

***

Анализируя работодателей мы выяснили, что компания Яндекс предлагает наибольшее количество вакансий и намного опережает все остальные компании. Так же мы узнали, что в России и Казахстане в общем пердставленны 410 и 207 работодателей соответственно, которые не имеют вакансий. Имея ввиду общее количество работодателей 23501, впринципе это не таккое большое количество, а конкретно по регионам России их ещё меньше, что даёт понять, что в каждом регионе компаний предоставляющих вакансиии намного больше. Компания Яндекс охватывает наибольшее количество регионов в которых предоставляет вакансии, далее следует Ростелеком и Спецремонт, затем количество регионов в которых представленны другие компании значительно сокращается. Так же мы выяснили, что у 8419 работодателей сфера деятельности не указанна, а это чуть больше трети всех представленных компаний. Примерно седьмая часть от общего числа представленных компаний указало своей сферой деятельности разработку ПО. Для компании Яндекс общее количество представленнх вакансий в городах миллиониках 485 из общего числа вакансий 1933 во всех регионах, что является четвертой частью. Следовательно в три четверти вакансий представленны в других менее населённых регионах.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [6]:
query_6_1 = f''' 
select
    count(v.id) cnt -- выбираем для вывода агрегацию счётчика вакансий по id
from vacancies v -- из таблицы вакансий с алиасом v
where lower(v.name) like '%data%' or lower(v.name) like '%данн%' -- задаём условие фильтрации, где ищем в столбце наименования, только те в которых встречаются такие последовательности как 'data' или 'данн'  
'''

In [7]:
df6_1 = pd.read_sql_query(query_6_1, connection) # создаём датафрейм из запроса
df6_1 # выводим датафрейм

C:\Users\usuario\AppData\Local\Temp\ipykernel_17968\812706.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df6_1 = pd.read_sql_query(query_6_1, connection)


cnt
0  1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [10]:
query_6_2 = f''' 
select 
    count(name) cnt_vacancies -- выбираем для вывода агрегацию счётчика наименования с алиасом cnt_vacancies
from vacancies -- из таблицы вакансий
where (name ilike '%data scientist%'
    or name ilike '%data science%'
    or name ilike '%исследователь данных%'
    or (name like '%ML%' and name not like '%HTML%')
    or name ilike '%machine learning%'
    or name ilike '%машинн%обучен%') -- задаём условия фильтрации наименования в столбце с наименованиями вакансий для выявления DS специализации
    and
    (name ilike '%junior%' 
    or experience ilike '%Нет опыта%'
    or employment ilike '%Стажировка%') -- добавляем доп. условия фильтрации для столбца опыта и типа трудоустройства для выявления начинающего специалиста
'''

In [11]:
df6_2 = pd.read_sql_query(query_6_2, connection) # создаём датафрейм из запроса
df6_2 # выводим датафрейм

C:\Users\usuario\AppData\Local\Temp\ipykernel_17968\3898269292.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df6_2 = pd.read_sql_query(query_6_2, connection)


cnt_vacancies
0             51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [12]:
query_6_3 = f''' 
select 
    count(name) -- выбираем для вывода агрегацию счётчика наименования
from vacancies -- из таблицы вакансий
where (name ilike '%data scientist%'
    or name ilike '%data science%'
    or name ilike '%исследователь данных%'
    or (name ilike '%ML%' and name not ilike '%html%')
    or name ilike '%machine learning%'
    or name ilike '%машинн%обучен%') -- задаём условия фильтрации наименования в столбце с наименованиями вакансий для выявления DS специализации
    and (key_skills ilike '%sql%' or key_skills ilike '%postgres%') -- добавляем условия фильтрации по ключевым навыкам
'''

In [13]:
df6_3 = pd.read_sql_query(query_6_3, connection) # создаём датафрейм из запроса
df6_3 # выводим датафрейм

C:\Users\usuario\AppData\Local\Temp\ipykernel_17968\1386111432.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df6_3 = pd.read_sql_query(query_6_3, connection)


count
0    229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [14]:
query_6_4 = f''' 
select 
    count(name) -- выбираем для вывода агрегацию счётчика наименования
from vacancies -- из таблицы вакансий
where (name ilike '%data scientist%'
    or name ilike '%data science%'
    or name ilike '%исследователь данных%'
    or (name ilike '%ML%' and name not ilike '%html%')
    or name ilike '%machine learning%'
    or name ilike '%машинн%обучен%') -- задаём условия фильтрации наименования в столбце с наименованиями вакансий для выявления DS специализации
    and (key_skills ilike '%python%') -- добавляем условия фильтрации по навыку Python
'''

In [15]:
df6_4 = pd.read_sql_query(query_6_4, connection)  # создаём датафрейм из запроса
df6_4 # выводим датафрейм

C:\Users\usuario\AppData\Local\Temp\ipykernel_17968\1196854611.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df6_4 = pd.read_sql_query(query_6_4, connection)


count
0    357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [15]:
query_6_5 = f''' 
select 
    round(avg(length(key_skills)-length(replace(key_skills, CHR(9), '')) +1), 2) cnt_skills -- задаём подсчёт округлённого до двух после запятой, среднего значения количества навыков указ. в вакансиях
from vacancies
where (name ilike '%data scientist%'
    or name ilike '%data science%'
    or name ilike '%исследователь данных%'
    or (name like '%ML%' and name not ilike '%html%')
    or name ilike '%machine learning%'
    or name ilike '%машинн%обучен%') -- задаём условия фильтрации наименования в столбце с наименованиями вакансий для выявления DS специализации
    and (key_skills is not null) -- указываем, что нам интересны только заполненные данные
'''

In [16]:
df6_5 = pd.read_sql_query(query_6_5, connection) # создаём датафрейм из запроса
df6_5 # выводим датафрейм

C:\Users\usuario\AppData\Local\Temp\ipykernel_21012\3037154696.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df6_5 = pd.read_sql_query(query_6_5, connection)


cnt_skills
0        6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [48]:
query_6_6 = f''' 
select 
    experience, -- выбираем для вывода столбец опыта
    round(avg(coalesce((salary_from+salary_to)/2, salary_from, salary_to))) -- высчитываем округлённую, среднюю зарплату
from vacancies -- из таблицы вакансий
where (name ilike '%data scientist%'
    or name ilike '%data science%'
    or name ilike '%исследователь данных%'
    or (name like '%ML%' and name not ilike '%html%')
    or name ilike '%machine learning%'
    or name ilike '%машинн%обучен%') -- задаём условия фильтрации наименования в столбце с наименованиями вакансий для выявления DS специализации
group by experience -- группируем по опыту
'''

In [49]:
df6_6 = pd.read_sql_query(query_6_6, connection) # создаём датафрейм из запроса
df6_6 # выводим датафрейм

C:\Users\usuario\AppData\Local\Temp\ipykernel_13628\2575030887.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df6_6 = pd.read_sql_query(query_6_6, connection) # создаём датафрейм из запроса


experience     round
0         Более 6 лет       NaN
1           Нет опыта   74643.0
2  От 1 года до 3 лет  139675.0
3       От 3 до 6 лет  243115.0

***

Изучачя вакансии по специализации DS мы выяснили, что всего представленных вакансий 1771. Для начинающих специалистов в сфере DS нашлась 51 вакансия. Для специалистов обладающими навыками в SQL и Postgres мы нашли 229 вакансий, а для специалистов обладающими навыками Python 357, что немного больше. В основном для специалистов в сфере DS запрашивают владением 6+ навыками. И специалисты в сфере DS могут расчитывать на зарплату 243 тыс. рублей в среднем с опытом от 3х до 6ти лет. Новички могут расчитывать на зарплату от 74 тыс. в среднем, а специалисты с опытом от года до 3х лет на зарплату 140 тыс в среднем.

# Общий вывод по проекту

У нас есть самое большое число вакансий (49197), предлагаемые 23501 работодателями, в 1362 х регионах и 294х сферах деятельности.
Болше всего вакансийв Москве и Санкт-Петербурге. Для более 50% от всего количества вакансий не указанна оплата. Средние значения минимального и предложения оплаты составили 71065 и 110537 соответственно. В большинстве случаев работодатели предпочитают полную занятость и полный день, а реже всего сменный график и проектную работу. Наиболее запрашиваемым является опыт от года до 3х лет, затем идёт опыт от 2х лет до 6ти. Наименьшее количество вакансий приходится, на тех у кого опыт более 6ти лет.
Компания Яндекс предлагает наибольшее количество вакансий. В России и Казахстане в общем пердставленны 410 и 207 работодателей соответственно, которые не имеют вакансий. Имея ввиду общее количество работодателей 23501, впринципе это не таккое большое количество, а конкретно по регионам России их ещё меньше, что даёт понять, что в каждом регионе компаний предоставляющих вакансиии намного больше. Компания Яндекс охватывает наибольшее количество регионов в которых предоставляет вакансии. У 8419 работодателей сфера деятельности не указанна, а это чуть больше трети всех представленных компаний. Примерно седьмая часть от общего числа представленных компаний указало своей сферой деятельности разработку ПО. Для компании Яндекс общее количество представленнх вакансий в городах миллиониках 485 из общего числа вакансий 1933 во всех регионах, что является четвертой частью. Следовательно в три четверти вакансий представленны в других менее населённых регионах.
Изучачя вакансии по специализации DS мы выяснили, что всего представленных вакансий 1771. Для начинающих специалистов в сфере DS нашлась 51 вакансия. Для специалистов обладающими навыками в SQL и Postgres мы нашли 229 вакансий, а для специалистов обладающими навыками Python 357, что немного больше. В основном для специалистов в сфере DS запрашивают владением 6+ навыками. И специалисты в сфере DS могут расчитывать на зарплату 243 тыс. рублей в среднем с опытом от 3х до 6ти лет. Новички могут расчитывать на зарплату от 74 тыс. в среднем, а специалисты с опытом от года до 3х лет на зарплату 140 тыс в среднем.

Учитывая предложения от работодателей в сфере DS, мы видим, что по большей части работодатели предпочитают брать на работу специалистов с опытом от года до 3х лет, обладающими 6ю ключевыми навыками, среди которых немалым спросом пользуются знание и умение работать с SQL и Python. Чуть меньше запрашиваются специалисты с опытом от 3х до 6ти лет и затем уже новички без опыта, конкретно в сфере DS для них нашлась 51 вакансия. Так же работодатели предпочитают нанимать специалистов на полный рабочий день и полную занятость. Больше всего вакансий в большем количестве регионов предоставляет компания Яндекс, что позволяет предположить, что специалист в сфере DS скорее сего найдёт себе работу в этой компании. Хотя надо учесть, что большой сегмент работодателей не указали свою сферу деятельности, а среди них могут оказатся такие, кто так же желал бы нанять специалистов в сфере DS.